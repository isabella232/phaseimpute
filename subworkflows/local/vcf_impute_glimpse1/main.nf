
include { GLIMPSE_PHASE                      } from '../../../modules/nf-core/glimpse/phase'
include { BCFTOOLS_INDEX as BCFTOOLS_INDEX_1 } from '../../../modules/nf-core/bcftools/index'
include { GLIMPSE_LIGATE                     } from '../../../modules/nf-core/glimpse/ligate'
include { BCFTOOLS_INDEX as BCFTOOLS_INDEX_2 } from '../../../modules/nf-core/bcftools/index'

workflow VCF_IMPUTE_GLIMPSE1 {

    take:
    ch_input        // channel (mandatory): [ [id], vcf, tbi ]
    ch_panel        // channel (mandatory): [ [panel, chr], vcf, tbi ]
    ch_chunks       // channel  (optional): [ [panel, chr], region1, region2 ]

    main:

    ch_versions = Channel.empty()
    ch_multiqc_files = Channel.empty()

    samples_file = Channel.of([[]]).collect()
    gmap_file    = Channel.of([[]]).collect()

    // Combine chunks with panel
    ch_chunks_panel = ch_chunks
        .combine(ch_panel, by:0)
        .map{ metaPC, regionin, regionout, panel, index ->
            [["panel": metaPC.id, "chr": metaPC.chr], regionin, regionout, panel, index]
        }

    // Join input and chunks reference
    ch_phase_input = ch_input
        .map{ metaIPC, vcf, index -> [metaIPC.subMap("panel", "chr"), metaIPC, vcf, index] }
        .combine(samples_file).view()
        .combine(ch_chunks_panel, by: 0).view()
        .combine(gmap_file)
        .map{ _metaPC, metaIPC, bam, bai, samples, regionin, regionout, panel, panel_index, gmap ->
            [metaIPC + ["chunk": regionout],
            bam, bai, samples, regionin, regionout, panel, panel_index, gmap]
        }

    GLIMPSE_PHASE ( ch_phase_input ) // [meta, vcf, index, sample, regionin, regionout, ref, ref_index, map]
    ch_versions = ch_versions.mix(GLIMPSE_PHASE.out.versions )

    BCFTOOLS_INDEX_1 ( GLIMPSE_PHASE.out.phased_variants )
    ch_versions = ch_versions.mix( BCFTOOLS_INDEX_1.out.versions )

    // Ligate all phased files in one and index it
    ligate_input = GLIMPSE_PHASE.out.phased_variants
        .join( BCFTOOLS_INDEX_1.out.csi )
        .map{ metaIPCR, vcf, index -> [metaIPCR.subMap("id", "panel", "chr", "batch"), vcf, index] }
        .groupTuple()

    GLIMPSE_LIGATE ( ligate_input )
    ch_versions = ch_versions.mix(GLIMPSE_LIGATE.out.versions )

    BCFTOOLS_INDEX_2 ( GLIMPSE_LIGATE.out.merged_variants )
    ch_versions = ch_versions.mix( BCFTOOLS_INDEX_2.out.versions )

    // Join imputed and index files
    ch_imputed_vcf_tbi = GLIMPSE_LIGATE.out.merged_variants
        .join(BCFTOOLS_INDEX_2.out.tbi)
        .map{ metaIPC, vcf, index -> [metaIPC + [tools: "glimpse1"], vcf, index] }

    emit:
    vcf_tbi             = ch_imputed_vcf_tbi    // channel: [ [id, panel, chr, tool], vcf, tbi ]
    versions            = ch_versions           // channel: [ versions.yml ]
    multiqc_files       = ch_multiqc_files      // channel: [ multiqc_files.yml ]
}
