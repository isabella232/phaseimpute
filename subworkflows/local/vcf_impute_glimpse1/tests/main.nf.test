nextflow_workflow {

    name "Test Subworkflow BAM_IMPUTE_GLIMPSE1"
    script "../main.nf"
    config "./nextflow.config"

    workflow "VCF_IMPUTE_GLIMPSE1"

    tag "subworkflows"
    tag "subworkflows_local"
    tag "subworkflows/vcf_impute_glimpse1"
    tag "vcf_impute_glimpse1"
    tag "bam_gl_bcftools"

    tag "glimpse"
    tag "glimpse/phase"
    tag "glimpse/ligate"
    tag "bcftools"
    tag "bcftools/index"

    test("Impute with glimpse1 two vcf") {
        when {
            workflow {
                """
                input[0] = Channel.fromList([
                    [
                        [id: "NA12878", panel: "1000GP", chr: "22"],
                        file(params.pipelines_testdata_base_path + "hum_data/individuals/NA12878/NA12878.s.bcf", checkIfExist:true),
                        file(params.pipelines_testdata_base_path + "hum_data/individuals/NA12878/NA12878.s.bcf.csi", checkIfExist:true),
                    ],
                    [
                        [id: "NA19401", panel: "1000GP", chr: "22"],
                        file(params.pipelines_testdata_base_path + "hum_data/individuals/NA19401/NA19401.s.bcf", checkIfExist:true),
                        file(params.pipelines_testdata_base_path + "hum_data/individuals/NA19401/NA19401.s.bcf.csi", checkIfExist:true),
                    ],
                ])
                input[1] = Channel.of([
                        [id: "1000GP", chr: "22"],
                        file(params.pipelines_testdata_base_path + "hum_data/panel/chr22/1000GP.chr22.s.norel.vcf.gz", checkIfExist:true),
                        file(params.pipelines_testdata_base_path + "hum_data/panel/chr22/1000GP.chr22.s.norel.vcf.gz.csi", checkIfExist:true),
                    ]).collect()
                input[2] = Channel.of(
                    [[chr: "22", id: "1000GP"], "chr22:16570065-16597215", "chr22:16570065-16592216"],
                    [[chr: "22", id: "1000GP"], "chr22:16587172-16609999", "chr22:16592229-16609999"]
                )
                """
            }
        }

        then {
            assertAll(
                { assert workflow.success },
                { assert snapshot(
                    workflow.out.versions,
                    workflow.out.vcf_tbi.collect{[
                        it[0],
                        path(it[1]).getFileName().toString(),
                        path(it[2]).getFileName().toString()
                    ] },
                    workflow.out.vcf_tbi.collect{ path(it[1]).vcf.summary },
                    workflow.out.vcf_tbi.collect{ path(it[1]).vcf.header.getGenotypeSamples().sort() }
                ).match() }
            )
        }
    }

    test("Impute with glimpse1 one vcf") {
        when {
            workflow {
                """
                input[0] = Channel.fromList([
                    [
                        [id: "NA12878", panel: "1000GP", chr: "22"],
                        file(params.pipelines_testdata_base_path + "hum_data/individuals/NA12878/NA12878.s.bcf", checkIfExist:true),
                        file(params.pipelines_testdata_base_path + "hum_data/individuals/NA12878/NA12878.s.bcf.csi", checkIfExist:true),
                    ]
                ])
                input[1] = Channel.of([
                        [id: "1000GP", chr: "22"],
                        file(params.pipelines_testdata_base_path + "hum_data/panel/chr22/1000GP.chr22.s.norel.vcf.gz", checkIfExist:true),
                        file(params.pipelines_testdata_base_path + "hum_data/panel/chr22/1000GP.chr22.s.norel.vcf.gz.csi", checkIfExist:true),
                    ]).collect()
                input[2] = Channel.of(
                    [[chr: "22", id: "1000GP"], "chr22:16570065-16597215", "chr22:16570065-16592216"],
                    [[chr: "22", id: "1000GP"], "chr22:16587172-16609999", "chr22:16592229-16609999"]
                )
                """
            }
        }

        then {
            assertAll(
                { assert workflow.success },
                { assert snapshot(
                    workflow.out.versions,
                    workflow.out.vcf_tbi.collect{[
                        it[0],
                        path(it[1]).getFileName().toString(),
                        path(it[2]).getFileName().toString()
                    ] },
                    workflow.out.vcf_tbi.collect{ path(it[1]).vcf.summary },
                    workflow.out.vcf_tbi.collect{ path(it[1]).vcf.header.getGenotypeSamples().sort() }
                ).match() }
            )
        }
    }
}
