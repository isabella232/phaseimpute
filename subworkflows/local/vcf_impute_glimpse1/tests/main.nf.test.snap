{
    "Impute with glimpse1 one vcf": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "NA12878",
                            "panel": "1000GP",
                            "chr": "22",
                            "tools": "glimpse1"
                        },
                        "NA12878_22_ligate.vcf.gz:md5,3a6cbabaafd2772afa140e02d3fe3238",
                        "NA12878_22_ligate.vcf.gz.tbi:md5,870a906079573453167617fc818638ab"
                    ]
                ],
                "1": [
                    "versions.yml:md5,2f8b6f8bb3b3972bb0375abf1de08807",
                    "versions.yml:md5,2f8b6f8bb3b3972bb0375abf1de08807",
                    "versions.yml:md5,8c79d92b93fe75a9094499ce5d5eaca2",
                    "versions.yml:md5,9d78fceae32ec42b1818bb5890e9a6ba",
                    "versions.yml:md5,bf4a479eb73467775a4afeebd25841aa",
                    "versions.yml:md5,bf4a479eb73467775a4afeebd25841aa"
                ],
                "2": [
                    
                ],
                "multiqc_files": [
                    
                ],
                "vcf_tbi": [
                    [
                        {
                            "id": "NA12878",
                            "panel": "1000GP",
                            "chr": "22",
                            "tools": "glimpse1"
                        },
                        "NA12878_22_ligate.vcf.gz:md5,3a6cbabaafd2772afa140e02d3fe3238",
                        "NA12878_22_ligate.vcf.gz.tbi:md5,870a906079573453167617fc818638ab"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,2f8b6f8bb3b3972bb0375abf1de08807",
                    "versions.yml:md5,2f8b6f8bb3b3972bb0375abf1de08807",
                    "versions.yml:md5,8c79d92b93fe75a9094499ce5d5eaca2",
                    "versions.yml:md5,9d78fceae32ec42b1818bb5890e9a6ba",
                    "versions.yml:md5,bf4a479eb73467775a4afeebd25841aa",
                    "versions.yml:md5,bf4a479eb73467775a4afeebd25841aa"
                ]
            },
            [
                "VcfFile [chromosomes=[chr22], sampleCount=1, variantCount=903, phased=false, phasedAutodetect=false]"
            ],
            [
                [
                    "NA12878"
                ]
            ]
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "24.10.0"
        },
        "timestamp": "2024-11-06T17:41:06.29924786"
    },
    "Impute with glimpse1 two vcf": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "NA12878",
                            "panel": "1000GP",
                            "chr": "22",
                            "tools": "glimpse1"
                        },
                        "NA12878_22_ligate.vcf.gz:md5,9dade6f748ddd821ec01775ad378ba42",
                        "NA12878_22_ligate.vcf.gz.tbi:md5,870a906079573453167617fc818638ab"
                    ],
                    [
                        {
                            "id": "NA19401",
                            "panel": "1000GP",
                            "chr": "22",
                            "tools": "glimpse1"
                        },
                        "NA19401_22_ligate.vcf.gz:md5,4c684a116c49da91d7271793f89f95e4",
                        "NA19401_22_ligate.vcf.gz.tbi:md5,7efba07d4dec3b525149f459c08d853b"
                    ]
                ],
                "1": [
                    "versions.yml:md5,2f8b6f8bb3b3972bb0375abf1de08807",
                    "versions.yml:md5,2f8b6f8bb3b3972bb0375abf1de08807",
                    "versions.yml:md5,2f8b6f8bb3b3972bb0375abf1de08807",
                    "versions.yml:md5,2f8b6f8bb3b3972bb0375abf1de08807",
                    "versions.yml:md5,8c79d92b93fe75a9094499ce5d5eaca2",
                    "versions.yml:md5,8c79d92b93fe75a9094499ce5d5eaca2",
                    "versions.yml:md5,9d78fceae32ec42b1818bb5890e9a6ba",
                    "versions.yml:md5,9d78fceae32ec42b1818bb5890e9a6ba",
                    "versions.yml:md5,bf4a479eb73467775a4afeebd25841aa",
                    "versions.yml:md5,bf4a479eb73467775a4afeebd25841aa",
                    "versions.yml:md5,bf4a479eb73467775a4afeebd25841aa",
                    "versions.yml:md5,bf4a479eb73467775a4afeebd25841aa"
                ],
                "2": [
                    
                ],
                "multiqc_files": [
                    
                ],
                "vcf_tbi": [
                    [
                        {
                            "id": "NA12878",
                            "panel": "1000GP",
                            "chr": "22",
                            "tools": "glimpse1"
                        },
                        "NA12878_22_ligate.vcf.gz:md5,9dade6f748ddd821ec01775ad378ba42",
                        "NA12878_22_ligate.vcf.gz.tbi:md5,870a906079573453167617fc818638ab"
                    ],
                    [
                        {
                            "id": "NA19401",
                            "panel": "1000GP",
                            "chr": "22",
                            "tools": "glimpse1"
                        },
                        "NA19401_22_ligate.vcf.gz:md5,4c684a116c49da91d7271793f89f95e4",
                        "NA19401_22_ligate.vcf.gz.tbi:md5,7efba07d4dec3b525149f459c08d853b"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,2f8b6f8bb3b3972bb0375abf1de08807",
                    "versions.yml:md5,2f8b6f8bb3b3972bb0375abf1de08807",
                    "versions.yml:md5,2f8b6f8bb3b3972bb0375abf1de08807",
                    "versions.yml:md5,2f8b6f8bb3b3972bb0375abf1de08807",
                    "versions.yml:md5,8c79d92b93fe75a9094499ce5d5eaca2",
                    "versions.yml:md5,8c79d92b93fe75a9094499ce5d5eaca2",
                    "versions.yml:md5,9d78fceae32ec42b1818bb5890e9a6ba",
                    "versions.yml:md5,9d78fceae32ec42b1818bb5890e9a6ba",
                    "versions.yml:md5,bf4a479eb73467775a4afeebd25841aa",
                    "versions.yml:md5,bf4a479eb73467775a4afeebd25841aa",
                    "versions.yml:md5,bf4a479eb73467775a4afeebd25841aa",
                    "versions.yml:md5,bf4a479eb73467775a4afeebd25841aa"
                ]
            },
            [
                "VcfFile [chromosomes=[chr22], sampleCount=1, variantCount=903, phased=false, phasedAutodetect=false]",
                "VcfFile [chromosomes=[chr22], sampleCount=1, variantCount=903, phased=false, phasedAutodetect=false]"
            ],
            [
                [
                    "NA12878"
                ],
                [
                    "NA19401"
                ]
            ]
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "24.10.0"
        },
        "timestamp": "2024-11-06T17:40:48.121360395"
    }
}